[WPE] WPE Platform: remove circular dependency between WPEView and WPEBuffer

carlosgcampos opened the pull request:
#### 5264807369a3d830ba359dd22fddad2a8d90fd73
<pre>
[WPE] WPE Platform: remove circular dependency between WPEView and WPEBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=293606">https://bugs.webkit.org/show_bug.cgi?id=293606</a>

Reviewed by Nikolas Zimmermann.

WPEBuffer keeps a reference to a WPEView and it&apos;s very common in WPEView
implementations to keep a reference of the buffer in use. The WPEBuffer
doesn&apos;t really need the view, in most of the cases it&apos;s only used to get
the display. So, we can change the view property by display and use a
weak reference instead.

Canonical link: <a href="https://commits.webkit.org/295435@main">https://commits.webkit.org/295435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee4afe0e077febacb1c953a79669754426b22977

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24839 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110341 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55799 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25265 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33383 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79844 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108131 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/19689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94885 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60151 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55182 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112879 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32290 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23779 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32656 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88551 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33443 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11231 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27696 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17040 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32213 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32003 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35346 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -160,7 +163,12 @@ static const struct wl_buffer_listener bufferListener = {
     {
         auto* buffer = WPE_BUFFER(userData);
-        wpe_view_buffer_released(wpe_buffer_get_view(buffer), buffer);
+        auto* dmaBufBuffer = static_cast<DMABufBuffer*>(wpe_buffer_get_user_data(buffer));
```
```
if (auto* dmaBufBuffer = static_cast<....>; dmaBuffer && dmaBuffer->view)
    wpe_view_buffer_released(...);
```

carlosgcampos replied:
I'm not a fan of that syntax, but I don't mind to change it either

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
```diff
@@ -94,11 +105,17 @@ static void wpeBufferDMABufDispose(GObject* object)
+    auto* display = wpe_buffer_get_display(buffer);
+    if (!display) {
+        priv->eglImage = nullptr;
+        g_set_error_literal(error, WPE_BUFFER_ERROR, WPE_BUFFER_ERROR_IMPORT_FAILED, "The WPE display of buffer has already been closed");
```
Is that the right terminology "has already been *closed*"?

carlosgcampos replied:
We can use disconnected if you prefer

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
```diff
@@ -64,21 +64,32 @@ struct _WPEBufferDMABufPrivate {
+    if (!display)
+        return;
+
+    if (auto* eglDisplay = wpe_display_get_egl_display(display, nullptr)) {
```
The second argument is a bit unfortunate, always have to check what that means..

carlosgcampos replied:
That's how errors work in GLib APIs

carlosgcampos committed:
[WPE] WPE Platform: remove circular dependency between WPEView and WPEBuffer
https://bugs.webkit.org/show_bug.cgi?id=293606

Reviewed by Nikolas Zimmermann.

WPEBuffer keeps a reference to a WPEView and it's very common in WPEView
implementations to keep a reference of the buffer in use. The WPEBuffer
doesn't really need the view, in most of the cases it's only used to get
the display. So, we can change the view property by display and use a
weak reference instead.

Canonical link: https://commits.webkit.org/295435@main
